fat_fs: Add reentrant zephyr support

Ox11 opened the pull request:
This commit enables zephyr to configure the FatFs FF_FS_REENTRANT
option and support fs actions from multiple threads.
CONFIG_FS_FATFS_REENTRANT enables the option and provides zephyr
mutex wrappers.

Signed-off-by: Nicola Ochsenbein <Nicola.Ochsenbein@husqvarnagroup.com>

de-nordic commented:
Missing west.yml manifest update to include change in external module https://github.com/zephyrproject-rtos/fatfs/pull/13/files:
https://docs.zephyrproject.org/latest/develop/modules.html#changes-to-existing-module

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -155,6 +155,11 @@ config FS_FATFS_WINDOW_ALIGNMENT
 	  that, in worst scenario, value provided here may cause FATFS
 	  structure to have size of twice the value
 
+config FS_FATFS_REENTRANT
```
This depends on posix subsystem.

de-nordic commented:
Is it possible to provide test scenarios, in Zephyr, to prove this works correctly?

Ox11 commented:
@de-nordic @carlescufi PR is updated. Could you comment?

de-nordic commented:
> @Ox11 thanks for the PR! After #51766 got merged this needs rework, can you please make the changes necessary? cc @de-nordic

There is also need to merge this https://github.com/zephyrproject-rtos/fatfs/pull/15 and west update here https://github.com/zephyrproject-rtos/zephyr/pull/52505, as the new FAT FS version brings fixes to reentrant code and also changes how synchronization primitives are used, so it directly affects this PR.

de-nordic reviewed (changes requested):
FIx compliance.
Separate commit for code and tests.
Missing commit with release notes update.

Ox11 commented:
@de-nordic I cant find the results of the compliance checks anymore. Is this because i force pushed the new changes out of habbit?
If so, could you please re-trigger them or help me find the old results :grimacing: ?

Ox11 committed:
modules: fs: Add reentrant zephyr support

This commit enables zephyr to configure the FatFs FF_FS_REENTRANT
option and support fs actions from multiple threads.
CONFIG_FS_FATFS_REENTRANT enables the option and provides zephyr
mutex wrappers.

Signed-off-by: Nicola Ochsenbein <Nicola.Ochsenbein@husqvarnagroup.com>

Ox11 committed:
test: fs: fat_fs_api: Add tests for reentrant zephyr support

This commit adds tests for the FatFs FF_FS_REENTRANT option.

Signed-off-by: Nicola Ochsenbein <Nicola.Ochsenbein@husqvarnagroup.com>

de-nordic commented on tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c:
```diff
@@ -0,0 +1,89 @@
+	/* File open should wait here, as the fs is locked. Therefore, automatic switch back to thread */
+	TC_PRINT("Open file\n");
+	res = fs_open(&filep, TEST_FILE, FS_O_CREATE | FS_O_RDWR);
+	zassert_false(res, "Err: File could not be opened [%d]\n", res);
```
`zasset_false` expects boolean; you probably would like to use `zassert_ok` here.

de-nordic commented on tests/subsys/fs/fat_fs_api/prj_reentrant.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_ZTEST_NEW_API=y
+CONFIG_FLASH=y
+CONFIG_FLASH_MAP=y
+CONFIG_FS_FATFS_REENTRANT=y
+CONFIG_MULTITHREADING=y
```
You do not have to make entire prj file to add these. The prj differs by two configs from the non reentrant one, so you can use `extra_config` option of testcases.yaml to run with original prj.conf.

de-nordic reviewed (changes requested):
1) fix compliance;
2) add yourself to ~~CODEOWNERS~~ MAINTAINERS for the zfs_ffsystem.c you have just added;
3) Add release notes that mentions adding reenatrans support for FATFS and addition of the Kconfig option CONFIG_FS_FATFS_REENTRANT.

Ox11 commented:
> 1. fix compliance;
> 2. add yourself to ~CODEOWNERS~ MAINTAINERS for the zfs_ffsystem.c you have just added;
> 3. Add release notes that mentions adding reenatrans support for FATFS and addition of the Kconfig option CONFIG_FS_FATFS_REENTRANT.

Compliance is fixed, I have added myself to the maintainers and added a release notes comment.

de-nordic commented on tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c:
```diff
@@ -0,0 +1,92 @@
+
+#include "test_fat.h"
+
+#if IS_ENABLED(CONFIG_FS_FATFS_REENTRANT)
```
Just use #ifdef. IS_ENABLED is for hacking `if` to work with `CONFIG_`s

de-nordic commented on tests/subsys/fs/fat_fs_api/src/test_fat.h:
```diff
@@ -1,13 +1,15 @@
 /*
  * Copyright (c) 2016 Intel Corporation.
  * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2022 Husqvarna AB
```
It is 23 now, took us long to review.

Ox11 replied:
Shall I also update the git commit date?

de-nordic commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -401,6 +401,7 @@ Libraries / Subsystems
   * Added new API call `fs_mkfs`.
   * Added new sample `samples/subsys/fs/format`.
   * FAT FS driver has been updated to version 0.15 w/patch1.
+  * Added FAT FS reentrant Kconfig :kconfig:option:`CONFIG_FS_FATFS_REENTRANT`.
```
The Kconfig is not reentrant. Rephrase this to be more clear that the Koconfig enables reentrant support in FAT FS.

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -211,6 +212,11 @@ config FS_FATFS_WINDOW_ALIGNMENT
 	  that, in worst scenario, value provided here may cause FATFS
 	  structure to have size of twice the value.
 
+config FS_FATFS_REENTRANT
+	bool "FatFs reentrant"
+	help
+	  Enable the re-entrancy (thread safe) option of the FatFs
+
```
Add more info on costs: for example that there will be mutex objects, one per volume. This is important because this takes some RAM, so people need to know.

de-nordic reviewed (changes requested):
Minor comments.

de-nordic commented on modules/fatfs/zfs_ffsystem.c:
```diff
@@ -0,0 +1,56 @@
+#endif /* FF_USE_LFN == 3 */
+
+#if FF_FS_REENTRANT	/* Mutual exclusion */
+static struct k_mutex Mutex[FF_VOLUMES + 1]; /* Table of Zephyr mutex */
```
Where the + 1 comes from?

Ox11 replied:
The + 1 is for the system mutex. Same as in the template file by ChaN. I added a comment about it.

nordicjm commented on modules/fatfs/CMakeLists.txt:
```diff
@@ -18,7 +18,10 @@ if(CONFIG_FAT_FILESYSTEM_ELM)
   zephyr_library_sources_ifdef(CONFIG_FS_FATFS_LFN
     ${ZEPHYR_FATFS_MODULE_DIR}/option/ffunicode.c
-    ${ZEPHYR_FATFS_MODULE_DIR}/option/ffsystem.c
+  )
+
+  zephyr_library_sources_ifdef((CONFIG_FS_FATFS_LFN OR CONFIG_FS_FATFS_REENTRANT)
+    zfs_ffsystem.c
```
This seems to be a wrong Kconfig order, if FS_FATFS_LFN mode requires reentrancy support, then it must pull FS_FATFS_REENTRANT in, either as a depends on or select, the this line is simplified

Ox11 replied:
They are not dependent on each other. Just both helper functions are implemented in the same file (same as in the example file provided by Elm Chan)
I did not get my local test running today, so I tried to test it directly on the pipeline :see_no_evil: 

Do you have a suggestion how I can include this zfs_ffsystem.c if either CONFIG_FS_FATFS_LFN or CONFIG_FS_FATFS_REENTRANT is set?

nordicjm replied:
This is OK then, I misread how it works.

nordicjm commented on modules/fatfs/zfs_ffsystem.c:
```diff
@@ -0,0 +1,69 @@
+/* Table of Zephyr mutex. One for each volume and an extra one for the ff system.
+ * See also the template file used as reference. Link is available in the header of this file.
+ */
+static struct k_mutex Mutex[FF_VOLUMES + 1];
```
snake case - lowercase characters

Ox11 commented:
@nordicjm @de-nordic Could you please re-trigger the pending workflows and re-review?

de-nordic commented:
The window for enhancements to 3.3 has been closed now, so moving this to 3.4.

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -510,6 +510,7 @@ Libraries / Subsystems
   * Added the option to disable CRC checking in :ref:`fcb_api` by enabling the
     Kconfig option :kconfig:option:`CONFIG_FCB_ALLOW_FIXED_ENDMARKER`
     and setting the `FCB_FLAGS_CRC_DISABLED` flag in the :c:struct:`fcb` struct.
+  * Added :kconfig:option:`CONFIG_FS_FATFS_REENTRANT` to enable the FAT FS reentrant option.
```
This will need to move to 3.4, but do that after 3.3 release is finished as the file will be created then

Ox11 replied:
Makes sense. However, do you know when 3.3 will be released and the 3.4 file created? - Doesn't this hold up all pending PRs?

nordicjm commented on modules/fatfs/zephyr_fatfs_config.h:
```diff
@@ -66,6 +67,14 @@
+#undef FF_FS_TIMEOUT
+#include <zephyr/kernel.h>
+#define FF_FS_REENTRANT		CONFIG_FS_FATFS_REENTRANT
+#define FF_FS_TIMEOUT		K_FOREVER
```
I'm not familiar with the internals of how this works, but let's assume that there is one partition with a log file and any other random file, if the log file is opened at startup and let's say it moves to a new file once an hour (or even once a day) but there's a way for files to be retrieved from this file system - what happens if a client requests the "other random file", does their file access part of their code (let's assume MCUmgr) then halt for **1 hour/1 day** until the mutex can be taken? If so, that seems rather inconvenient for all users of this, and unacceptable

Ox11 replied:
The reentrant option does not lock the media upon open. It only locks file access operations so that volume access is serialized. E.g. not two threads are trying to send commands over the SPI to the external Flash/SD. You only need to wait for the other open/write/... operation is finsihed.
I added a test that demonstrates that you may still have two files opened at the same time accessing them from different threads. Hope that helps.

de-nordic replied:
> I'm not familiar with the internals of how this works, but let's assume that there is one partition with a log file and any other random file, if the log file is opened at startup and let's say it moves to a new file once an hour (or even once a day) but there's a way for files to be retrieved from this file system - what happens if a client requests the "other random file", does their file access part of their code (let's assume MCUmgr) then halt for **1 hour/1 day** until the mutex can be taken? If so, that seems rather inconvenient for all users of this, and unacceptable

This will not be a problem. Generally most of functions, there are exceptions, call `validate` (https://github.com/zephyrproject-rtos/fatfs/blob/master/ff.c#L3613) that does locking for the operation, and do `LEAVE_FF` to retrun that does unlocking, https://github.com/zephyrproject-rtos/fatfs/blob/427159bf95ea49b7680facffaa29ad506b42709b/ff.c#L237.

arbrauns commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -211,6 +212,13 @@ config FS_FATFS_WINDOW_ALIGNMENT
 
+config FS_FATFS_REENTRANT
+	bool "FatFs reentrant"
+	help
```
Would a `depends on !FS_FATFS_LFN_MODE_BSS` make sense here? Without, it's really easy to enable LFN in its default configuration, enable reentrancy, and not realize that it's actually still not thread safe. With the `depends`, there would at least be a warning because `CONFIG_FS_FATFS_REENTRANT` can't be selected.

de-nordic replied:
Yeah, we should probably add that.

Ox11 replied:
Good point!
I tested it locally with `depends on !FS_FATFS_LFN_MODE_BSS`:
I added the `CONFIG_FS_FATFS_LFN=y` in the testcase.yaml for the reentrant test in the `extra_configs`, but the test did not fail until I set the mode manually with `FS_FATFS_LFN_MODE_BSS=y`. I also tested it by adding the `depends on !FS_FATFS_REENTRANT` to `FS_FATFS_LFN_MODE_BSS` and had the same behaviour.
The default mode is probably the most critical choise to warn about which still won't work.
Do you know more about this behaviour?

Ox11 replied:
I added `depends on !FS_FATFS_REENTRANT` to `FS_FATFS_LFN_MODE_BSS` because `FS_FATFS_LFN_MODE_BSS` is not always visible and therefore I think this way round it is cleaner.

nordicjm reviewed (commented):
this can rebased now to use the 3.4 release note file

Ox11 committed:
MAINTAINERS: Add ox11 to FatFs

ox11 is maintaining on FatFs.

Signed-off-by: Nicola Ochsenbein <Nicola.Ochsenbein@husqvarnagroup.com>

Ox11 commented on MAINTAINERS.yml:
```diff
@@ -1240,6 +1240,16 @@ Filesystems:
   labels:
     - "area: File System"
 
+"Filesystems: FatFs":
```
Done.
But I am not quite sure if i should add me as a maintainer because I feel like I am just a collaborator. But I don't know if I can add a subsystem entry without a maintainer. Any ideas? @de-nordic @nordicjm

Ox11 commented:
> this can rebased now to use the 3.4 release note file

Done.

Ox11 committed:
doc: release-notes: Update release notes 3.4

Update release notes 3.4 with the new FatFs reentrant Kconfig option.

Signed-off-by: Nicola Ochsenbein <Nicola.Ochsenbein@husqvarnagroup.com>

de-nordic commented on MAINTAINERS.yml:
```diff
@@ -1347,6 +1347,16 @@ Filesystems:
   labels:
     - "area: File System"
 
+"Filesystems: FatFs":
```
Rename this as "FatFs reentrant support"

Ox11 replied:
Done.

de-nordic commented on MAINTAINERS.yml:
```diff
@@ -1347,6 +1347,16 @@ Filesystems:
+  maintainers:
+    - ox11
+  files:
+    - subsys/fs/fat_fs.c
```
I see no reason to pass Maintainership on this file, it is not even touched in this PR. Remove it from list.

Ox11 replied:
Yes, you are right.

de-nordic commented on MAINTAINERS.yml:
```diff
@@ -1347,6 +1347,16 @@ Filesystems:
+    - ox11
+  files:
+    - subsys/fs/fat_fs.c
+    - modules/fatfs/zfs_ffsystem.c
```
List is missing test_fat_file_reentrant.c file.

Ox11 replied:
Added it.

Ox11 commented:
Could you please review this PR again @nordicjm @de-nordic ? Did I miss something that I have to do?

Ox11 commented:
Thank you for your approves @nordicjm @de-nordic. Who can/will merge this PR? - I canot merge it, as I am not authorized.

stephanosio commented on MAINTAINERS.yml:
```diff
@@ -1371,6 +1371,16 @@ Filesystems:
   labels:
     - "area: File System"
 
+"Filesystems: FatFs reentrant support":
+  status: maintained
+  maintainers:
+    - ox11
+  files:
+    - modules/fatfs/zfs_ffsystem.c
+    - tests/subsys/fs/fat_fs_api/src/test_fat_file_reentrant.c
+  labels:
+    - "area: File System"
```
[From the documentation:](https://docs.zephyrproject.org/latest/project/project_roles.html#maintainers-file)

> We should keep the granularity of code maintainership at a manageable level

It does not make sense to have "FatFs reentrant support" as a dedicated area -- this is too much granularity.

Please revert this change or define the maintainership as "Filesystems: Fatfs."

Ox11 replied:
I am not a maintainer of the entire FatFS module.
My suggestion would be to either create a new module like this
```
"Filesystems: FatFs":
  status: maintained
  maintainers:
    - de-nordic
  collaborators:
    - ox11
  files:
    - modules/fatfs
    - tests/subsys/fs/common
    - tests/subsys/fs/fat_fs_api
    - tests/subsys/fs/fat_fs_dual_drive
  labels:
    - "area: File System"
```

or to add myself to the collaborators in the fs area
```
Filesystems:
  status: maintained
  maintainers:
    - de-nordic
  collaborators:
    - Laczen
    - nashif
    - ox11
  files:
    - include/zephyr/fs/
    - samples/subsys/fs/
    - subsys/fs/
    - tests/subsys/fs/
  labels:
    - "area: File System"
```

The first option should IMHO be created by @de-nordic as he would be mentioned as maintainer. Additionally, is there anyone else that should be mentioned as collaborator?
